File: ui/demo/record_video.py
#!/usr/bin/env python3
"""
RoamEN UI Demo Video Recorder (Python version)
Records the demo UI in action using Playwright

Requirements:
    pip install playwright
    playwright install chromium
"""

import asyncio
import os
from pathlib import Path
from playwright.async_api import async_playwright


class DemoRecorder:
    def __init__(self):
        self.width = 1920
        self.height = 1080
        self.fps = 30
        self.output_file = "roamen-demo.webm"

    async def record(self):
        print("🎬 Starting RoamEN UI Demo Recording...")
        print(f"📐 Resolution: {self.width}x{self.height}")
        print(f"🎞️  Frame rate: {self.fps} fps")

        async with async_playwright() as p:
            # Launch browser
            print("🌐 Launching browser...")
            browser = await p.chromium.launch(
                headless=True,
                args=[
                    '--no-sandbox',
                    '--disable-setuid-sandbox',
                    '--disable-dev-shm-usage',
                ]
            )

            # Create context with video recording
            context = await browser.new_context(
                viewport={'width': self.width, 'height': self.height},
                record_video_dir='.',
                record_video_size={'width': self.width, 'height': self.height}
            )

            page = await context.new_page()

            # Navigate to demo page
            html_path = Path(__file__).parent / 'index.html'
            print(f"📄 Loading UI from: {html_path}")
            await page.goto(f'file://{html_path.absolute()}')

            # Wait for page to load
            await page.wait_for_timeout(2000)

            print("🔴 Recording started...")

            # Record for 57 seconds (demo duration + buffer)
            recording_duration = 57000
            print(f"⏱️  Recording for {recording_duration / 1000} seconds...")

            # Show progress
            for i in range(0, recording_duration, 1000):
                await page.wait_for_timeout(1000)
                progress = min(100, (i / recording_duration) * 100)
                print(f"\r⏳ Progress: {progress:.1f}%", end='', flush=True)

            print("\r⏳ Progress: 100.0%")

            # Close context to save video
            print("⏹️  Stopping recording...")
            await context.close()
            await browser.close()

            # Rename video file
            video_path = await page.video.path()
            if video_path:
                output_path = Path(__file__).parent / self.output_file
                os.rename(video_path, output_path)
                print(f"✅ Recording complete!")
                print(f"📹 Video saved to: {output_path}")

                # Get file size
                file_size_mb = output_path.stat().st_size / (1024 * 1024)
                print(f"📦 File size: {file_size_mb:.2f} MB")
            else:
                print("❌ Video recording failed")


async def main():
    recorder = DemoRecorder()
    try:
        await recorder.record()
        print("🎉 Done!")
    except Exception as e:
        print(f"❌ Error during recording: {e}")
        raise


if __name__ == "__main__":
    asyncio.run(main())

File: ui/demo/record-video.js
#!/usr/bin/env node

/**
 * RoamEN UI Demo Video Recorder
 * Records the demo UI in action and generates a video file
 *
 * Requirements: puppeteer, puppeteer-screen-recorder
 * Install: npm install puppeteer puppeteer-screen-recorder
 */

const puppeteer = require('puppeteer');
const { PuppeteerScreenRecorder } = require('puppeteer-screen-recorder');
const path = require('path');
const fs = require('fs');

const CONFIG = {
    width: 1920,
    height: 1080,
    fps: 30,
    videoBitrate: 5000,
    videoCodec: 'libx264',
    videoPreset: 'medium',
    videoCrf: 18,
    aspectRatio: '16:9'
};

async function recordDemo() {
    console.log('🎬 Starting RoamEN UI Demo Recording...');
    console.log('📐 Resolution:', `${CONFIG.width}x${CONFIG.height}`);
    console.log('🎞️  Frame rate:', `${CONFIG.fps} fps`);

    let browser;
    let recorder;

    try {
        // Launch browser
        console.log('🌐 Launching browser...');
        browser = await puppeteer.launch({
            headless: true,
            args: [
                '--no-sandbox',
                '--disable-setuid-sandbox',
                '--disable-dev-shm-usage',
                '--disable-gpu',
                `--window-size=${CONFIG.width},${CONFIG.height}`
            ]
        });

        const page = await browser.newPage();

        // Set viewport
        await page.setViewport({
            width: CONFIG.width,
            height: CONFIG.height,
            deviceScaleFactor: 1
        });

        // Setup screen recorder
        recorder = new PuppeteerScreenRecorder(page, {
            followNewTab: false,
            fps: CONFIG.fps,
            ffmpeg_Path: null, // Uses system ffmpeg
            videoFrame: {
                width: CONFIG.width,
                height: CONFIG.height
            },
            aspectRatio: CONFIG.aspectRatio,
            videoCrf: CONFIG.videoCrf,
            videoCodec: CONFIG.videoCodec,
            videoPreset: CONFIG.videoPreset,
            videoBitrate: CONFIG.videoBitrate
        });

        // Navigate to the demo page
        const htmlPath = path.join(__dirname, 'index.html');
        console.log('📄 Loading UI from:', htmlPath);
        await page.goto(`file://${htmlPath}`, { waitUntil: 'networkidle0' });

        // Wait for page to be fully loaded
        await page.waitForTimeout(2000);

        // Start recording
        const outputPath = path.join(__dirname, 'roamen-demo.mp4');
        console.log('🔴 Recording started...');
        console.log('💾 Output file:', outputPath);

        await recorder.start(outputPath);

        // Wait for demo to complete (52 seconds + 5 second buffer)
        const recordingDuration = 57000;
        console.log(`⏱️  Recording for ${recordingDuration / 1000} seconds...`);

        // Show progress
        const progressInterval = setInterval(() => {
            const elapsed = Date.now() - startTime;
            const percentage = Math.min(100, (elapsed / recordingDuration) * 100);
            process.stdout.write(`\r⏳ Progress: ${percentage.toFixed(1)}%`);
        }, 1000);

        const startTime = Date.now();
        await page.waitForTimeout(recordingDuration);

        clearInterval(progressInterval);
        process.stdout.write('\r⏳ Progress: 100.0%\n');

        // Stop recording
        console.log('⏹️  Stopping recording...');
        await recorder.stop();

        console.log('✅ Recording complete!');
        console.log('📹 Video saved to:', outputPath);

        // Get file size
        const stats = fs.statSync(outputPath);
        const fileSizeMB = (stats.size / (1024 * 1024)).toFixed(2);
        console.log('📦 File size:', `${fileSizeMB} MB`);

    } catch (error) {
        console.error('❌ Error during recording:', error);
        throw error;
    } finally {
        // Cleanup
        if (recorder) {
            try {
                await recorder.stop();
            } catch (e) {
                // Ignore errors during cleanup
            }
        }
        if (browser) {
            await browser.close();
        }
    }
}

// Main execution
if (require.main === module) {
    recordDemo()
        .then(() => {
            console.log('🎉 Done!');
            process.exit(0);
        })
        .catch(error => {
            console.error('💥 Fatal error:', error);
            process.exit(1);
        });
}

module.exports = { recordDemo };
